src/concurrent.rs

```diff
@@ -0,0 +1,170 @@
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use nix::unistd;
+use std::fs;
+use std::io;
+use std::os::unix::io as unix_io;
+use std::sync::{Arc, Mutex};
+
+/// A file with a single scoped owner but with multiple non-owner views.
+///
+/// A `ShareableFile` object owns the file passed to it at construction time and will close the
+/// underlying file handle when the object is dropped.
+///
+/// Concurrent views into this same file, obtained via the `clone_unowned` method, do not own the
+/// file handle.  Such views must accept the fact that the handle can be closed at any time, a
+/// condition that is simply exposed as if the handle reached EOF.  Concurrent views can safely be
+/// moved across threads.
+pub struct ShareableFile {
+    /// Underlying file descriptor shared across all views of this file.
+    fd: unix_io::RawFd,
+
+    /// Whether this instance of the file is responsible for closing the file descriptor.
+    owned: bool,
+
+    /// Whether the file descriptor has already been closed or not.
+    ///
+    /// In principle, we don't need this field: concurrent readers of this file will get their
+    /// reads abruptly terminated with an error, and we should be able to rely on EBADFD to tell
+    /// that this happened because of us closing the file descriptor.  Unfortunately... macOS
+    /// Mojave doesn't cooperate and yields very strange error codes (like EISDIR) pretty frequently
+    /// (but not deterministically).  May it be an APFS bug?
+    closed: Arc<Mutex<bool>>,
+}
+
+impl ShareableFile {
+    /// Constructs a new `ShareableFile` from an open file and takes ownership of it.
+    pub fn from(file: fs::File) -> ShareableFile {
+        use std::os::unix::io::IntoRawFd;
+        ShareableFile {
+            fd: file.into_raw_fd(),
+            owned: true,
+            closed: Arc::from(Mutex::from(false)),
+        }
+    }
+
+    /// Returns an unowned view of the file.
+    ///
+    /// Users of this file must accept that the file can be closed at any time by the owner.
+    pub fn clone_unowned(&mut self) -> ShareableFile {
+        ShareableFile {
+            fd: self.fd,
+            owned: false,
+            closed: self.closed.clone(),
+        }
+    }
+}
+
+impl Drop for ShareableFile {
+    fn drop(&mut self) -> () {
+        if self.owned {
+            debug!("Closing ShareableFile with fd {}", self.fd);
+
+            // We must mark the file as closed before attempting to close it because, when we do
+            // the actual close, any threads blocked on a read will get unlocked immediately and
+            // we won't have a chance to update the status.  It doesn't matter if the close succeeds
+            // or not though, because we cannot do anything useful in the latter case anyway.
+            *self.closed.lock().unwrap() = true;
+
+            match unistd::close(self.fd) {
+                Ok(_) => (),
```

Comment:
```suggestion
                Ok(()) => (),
```
I prefer this because when I see `_` I wonder what you are ignoring.